src/tree-sitter-language-mode.js

```diff
@@ -384,11 +364,273 @@ class TreeSitterLanguageMode {
   getGrammar () {
     return this.grammar
   }
+
+  /*
+  Section - Private
+  */
+
+  grammarForLanguageString (languageString) {
+    return this.grammarRegistry.treeSitterGrammarForLanguageString(languageString)
+  }
+
+  emitRangeUpdate (range) {
+    const startRow = range.start.row
+    const endRow = range.end.row
+    for (let row = startRow; row < endRow; row++) {
+      this.isFoldableCache[row] = undefined
+    }
+    this.emitter.emit('did-change-highlighting', range)
+  }
 }
 
-class TreeSitterHighlightIterator {
-  constructor (languageMode, treeCursor) {
+class LanguageLayer {
+  constructor (languageMode, grammar) {
     this.languageMode = languageMode
+    this.grammar = grammar
+    this.tree = null
+    this.currentParsePromise = null
+    this.patchSinceCurrentParseStarted = null
+  }
+
+  buildHighlightIterator () {
+    if (this.tree) {
+      return new LayerHighlightIterator(this, this.tree.walk())
+    } else {
+      return new NullHighlightIterator()
+    }
+  }
+
+  handleTextChange ({oldRange, newRange, oldText, newText}) {
+    if (this.tree) {
+      this.tree.edit(this._treeEditForBufferChange(
+        oldRange.start, oldRange.end, newRange.end, oldText, newText
+      ))
+    }
+
+    if (this.currentParsePromise) {
+      if (!this.patchSinceCurrentParseStarted) {
+        this.patchSinceCurrentParseStarted = new Patch()
+      }
+      this.patchSinceCurrentParseStarted.splice(
+        oldRange.start,
+        oldRange.end,
+        newRange.end,
+        oldText,
+        newText
+      )
+    }
+  }
+
+  destroy() {
+    for (const marker of this.languageMode.injectionsMarkerLayer.getMarkers()) {
+      if (marker.parentLanguageLayer === this) {
+        marker.languageLayer.destroy()
+        marker.destroy()
+      }
+    }
+  }
+
+  async update (containingNode) {
+    if (this.currentParsePromise) return this.currentParsePromise
+
+    this.currentParsePromise = this._performUpdate(containingNode)
+    await this.currentParsePromise
+    this.currentParsePromise = null
+
+    if (this.patchSinceCurrentParseStarted) {
+      const changes = this.patchSinceCurrentParseStarted.getChanges()
+      for (let i = changes.length; i --> 0;) {
+        const {oldStart, oldEnd, newEnd, oldText, newText} = changes[i]
+        this.tree.edit(this._treeEditForBufferChange(
+          oldStart, oldEnd, newEnd, oldText, newText
+        ))
+      }
+      this.patchSinceCurrentParseStarted = null
+      this.update(containingNode)
+    }
+  }
+
+  async _performUpdate (containingNode) {
```

Comment:
Stupid style comment; any specific reason for prefixing some of these functions with an underscore? I'm assuming they are "private", but I don't think we do this anywhere else in the core code?